sled-agent/src/http_entrypoints.rs

```diff
@@ -64,9 +64,28 @@ async fn services_put(
     rqctx: RequestContext<SledAgent>,
     body: TypedBody<ServiceEnsureBody>,
 ) -> Result<HttpResponseUpdatedNoContent, HttpError> {
-    let sa = rqctx.context();
+    let sa = rqctx.context().clone();
     let body_args = body.into_inner();
-    sa.services_ensure(body_args).await.map_err(|e| Error::from(e))?;
+
+    // Spawn a task to run `services_ensure`: dropshot/hyper will cancel an
+    // endpoint's task if the call times out or is otherwise cancelled, and this
+    // leads to a specific issue where booting all the service zones takes
+    // longer than the progenitor client default timeout. The request times out,
+    // `services_ensure` gets cancelled, and this leaves zones partially
+    // configured but not added to the list of `existing_zones`. When the next
+    // `PUT /services` call is made, `services_ensure` will eventually try to
+    // bring up the same zone it was interrupted at, leading to configuration
+    // issues. See: oxidecomputer/omicron#3098.
+
+    match tokio::spawn(async move { sa.services_ensure(body_args).await }).await
```

Comment:
Ideally it would seem good to have a limit to how many of these tasks can be outstanding.  I'm not sure this is worth prioritizing now though.

Reply:
I thought about changing where `services_ensure` grabs the lock from a `lock` to a `try_lock`, bubbling up the `TryLockError` as maybe a 429. This means that each client will now have to catch this error and retry on timeout and this new 429. That's not a terrible amount of work but it does mean that all current and future callers will have to do this.